cmd/nerdctl/cp.go

```diff
@@ -0,0 +1,609 @@
+
+	// Init tar Pipe is the first step to process copy to container
+	t := initTarPipe(&copyConfig)
+
+	if _, err := os.Stat(t.src); err != nil {
+		return fmt.Errorf("%s doesn't exist in local filesystem", t.src)
+	}
+
+	// TarPipe destination should exist in the container and it should be a directory
+	destination, err := securejoin.SecureJoin(filepath.Join("/proc/", fmt.Sprint(t.option.containerPid), "/root"), t.dst)
+	if err != nil {
+		return err
+	}
+
+	d, err := os.Stat(destination)
+	if err != nil {
+		return fmt.Errorf("%s should exists", t.dst)
+	} else {
+		if !d.IsDir() {
+			return fmt.Errorf("%s should be a directory", t.dst)
+		}
+	}
+
+	go func(*TarPipe) {
+		defer t.outStream.Close()
+		makeTar(cmd, t)
+	}(t)
+
+	// check if tar binary exists in the container
+	tarBinary, err := copyutil.TarBinary()
+	if err != nil {
+		hint := fmt.Sprintf("`tar` binary needs to be installed")
+		return fmt.Errorf(hint+": %w", err)
+	}
+
+	var tarArgs []string
+	tarArgs = []string{"-xmf", "-"}
+
+	dstDir, _ := copyutil.SplitPathDirEntry(filepath.Clean(t.dst))
+	var target string
+	if len(dstDir) > 0 {
+		target, err = securejoin.SecureJoin(filepath.Join("/proc/", fmt.Sprint(t.option.containerPid), "/root"), dstDir)
+		if err != nil {
+			return err
+		}
+	} else {
+		target, err = securejoin.SecureJoin(filepath.Join("/proc/", fmt.Sprint(t.option.containerPid), "/root"), "")
+		if err != nil {
+			return err
+		}
+	}
+
+	tarArgs = append(tarArgs, "-C", target)
+	command := exec.Command(tarBinary, tarArgs...)
+	stdin, err := command.StdinPipe()
+	if err != nil {
+		return err
+	}
+
+	go func(*TarPipe) error {
+
+		defer stdin.Close()
+		isquiet, err := cmd.Flags().GetBool("quiet")
+		if err != nil {
+			return err
+		}
+
+		var writer io.Writer
+		writer = stdin
+
+		if !isquiet {
+			bar := progressbar.DefaultBytes(
+				-1,
+				"uploading",
+			)
+
+			writer = io.MultiWriter(stdin, bar)
+		}
+
+		if _, err := io.Copy(writer, t.reader); err != nil {
+			return err
+		}
+
+		return nil
+	}(t)
+
+	out, err := command.CombinedOutput()
+	if err != nil {
+		logrus.Errorf("failed to exec: %v: %v", string(out), err)
+	}
+
+	return nil
+}
+
+func makeTar(cmd *cobra.Command, t *TarPipe) error {
+
+	// making Tar flow:
+	// tar --> buf
+
+	tw := tar.NewWriter(t.outStream)
+	defer tw.Close()
+
+	srcDir, srcBase := copyutil.SplitPathDirEntry(filepath.Clean(t.src))
+	dstDir, dstBase := copyutil.SplitPathDirEntry(filepath.Clean(t.dst))
+
+	// At this point, we consider that destination exists and it is a directory
+	// Override dstBase by copying specified src into it
+	dstBase = filepath.Join(dstBase, srcBase)
+
+	archive, err := cmd.Flags().GetBool("archive")
+	if err != nil {
+		return err
+	}
+
+	return recursiveTar(srcDir, srcBase, dstDir, dstBase, t.option.containerUser, tw, archive)
+}
+
+// recursiveTar makes a tar recursively
+func recursiveTar(srcDir, srcBase, dstDir, dstBase string, user runtimespec.User, tw *tar.Writer, archive bool) error {
```

Comment:
Just using `tar` command on the host would be simplier

Reply:
@AkihiroSuda  I need `archive/tar` because I writing to `io.PipeWriter` here  https://github.com/containerd/nerdctl/pull/643/files#diff-dc36d8c99566d6fecdc69b09e1781108cb3026a4a5dbbfe31e1df4b4d2572c3bR331 .  I could have used archive/tar everywhere but it can be a possible evolution (future PR) to avoid a vulnerable binary tar